ivadomed/loader/utils.py

```diff
@@ -722,149 +722,221 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
-    # TODO: verify merge behavior with EEG and DWI scans files, tested with anat and microscopy only
-    df_scans = pd.DataFrame()
-    for root, dirs, files in os.walk(bids_path):
-        for file in files:
-            if file.endswith("scans.tsv"):
-                df_temp = pd.read_csv(os.path.join(root, file), sep='\t')
-                df_scans = pd.concat([df_scans, df_temp], ignore_index=True)
-    if not df_scans.empty:
-        df_scans['filename'] = df_scans['filename'].apply(os.path.basename)
-        df = pd.merge(df, df_scans, on=['filename'], suffixes=("_x", None), how='left')
-
-    # TODO: check if other files are needed for EEG and DWI
-
-    # If indexing of derivatives is true
-    # Get list of subject files with available derivatives
-    if derivatives:
-        subject_files = df[~df['path'].str.contains('derivatives')]['filename'].to_list()
-        prefix_fnames = []
-        [prefix_fnames.append(s.split('.')[0]) for s in subject_files]
-        deriv = df[df['path'].str.contains('derivatives')]['filename'].tolist()
-        has_deriv = []
-        if roi_suffix is not None:
-            for p in prefix_fnames:
-                available = [d for d in deriv if p in d]
-                if roi_suffix in ('|'.join(available)):
-                    has_deriv.append(p)
-                else:
-                    logger.warning("Missing ROI derivatives for subject {}. Skipping subject.".format(p))
-        else:
-            for p in prefix_fnames:
-                available = [d for d in deriv if p in d]
+    def __init__(self, loader_params, derivatives, log_directory):
+
+        # bids_path from loader parameters
+        self.bids_path = os.path.join(loader_params['bids_path'], '')
+
+        # bids_config from loader parameters
+        self.bids_config = None if 'bids_config' not in loader_params else loader_params['bids_config']
+
+        # target_suffix and roi_suffix from loader parameters
+        self.target_suffix = loader_params['target_suffix']
+        # If `target_suffix` is a list of lists convert to list
+        if any(isinstance(t, list) for t in self.target_suffix):
+            self.target_suffix = list(itertools.chain.from_iterable(self.target_suffix))
+        self.roi_suffix = loader_params['roi_params']['suffix']
+        # If `roi_suffix` is not None, add to target_suffix
+        if self.roi_suffix is not None:
+            self.target_suffix.append(self.roi_suffix)
+
+        # extensions from loader parameters
+        self.extensions = loader_params['extensions']
+
+        # contrast_lst from loader parameters
+        self.contrast_lst = loader_params["contrast_params"]["contrast_lst"]
+
+        # derivatives
+        self.derivatives = derivatives
+
+        # Create dataframe
+        self.create_bids_dataframe()
+
+        # Save dataframe as csv file
+        self.save(os.path.join(log_directory, "bids_dataframe.csv"))
+
+    def create_bids_dataframe(self):
```

Comment:
Overall, really great work! I do have a suggestion for this function. Since it is over 100 lines long, it may be worth the effort to extract some of the nested conditions into helper functions. This long-term would make maintenance of the class less costly and also reduce some of complexity and readability.